linode/token/framework_resource.go

```diff
@@ -0,0 +1,270 @@
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	expireStr := data.Expiry.ValueString()
+	dt, err := time.Parse(time.RFC3339, expireStr)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Invalid datetime string",
+			fmt.Sprintf(
+				"Expected expiry to be an time.RFC3339 datetime string (e.g., %s), got %s",
+				time.RFC3339,
+				expireStr,
+			),
+		)
+		return
+	}
+
+	createOpts := linodego.TokenCreateOptions{
+		Label:  data.Label.ValueString(),
+		Scopes: data.Scopes.ValueString(),
+		Expiry: &dt,
+	}
+
+	token, err := client.CreateToken(ctx, createOpts)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Token creation error",
+			err.Error(),
+		)
+		return
+	}
+
+	data.parseToken(token)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *Resource) Read(
+	ctx context.Context,
+	req resource.ReadRequest,
+	resp *resource.ReadResponse,
+) {
+	client := r.client
+
+	var data ResourceModel
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	id := helper.StringToInt64(data.ID.ValueString(), resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	token, err := client.GetToken(ctx, int(id))
+	if err != nil {
+		if lerr, ok := err.(*linodego.Error); ok && lerr.Code == 404 {
+			resp.Diagnostics.AddWarning(
+				"Token No Longer Exists",
+				fmt.Sprintf(
+					"Removing Linode Token with ID %v from state because it no longer exists",
+					data.ID,
+				),
+			)
+			resp.State.RemoveResource(ctx)
+			return
+		}
+		resp.Diagnostics.AddError(
+			"Unable to Refresh the Token",
+			fmt.Sprintf(
+				"Error finding the specified Linode Token: %s",
+				err.Error(),
+			),
+		)
+	}
+
+	data.parseToken(token)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *Resource) Update(
+	ctx context.Context,
+	req resource.UpdateRequest,
+	resp *resource.UpdateResponse,
+) {
+	var data ResourceModel
+	var tokenIDString string
+	resp.Diagnostics.Append(
+		req.State.GetAttribute(ctx, path.Root("id"), &tokenIDString)...,
+	)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tokenID := int(helper.StringToInt64(tokenIDString, resp.Diagnostics))
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	client := r.client
+	token, err := client.GetToken(ctx, tokenID)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("Failed to get the token with id %v", tokenID),
+			err.Error(),
+		)
+		return
+	}
+
+	updateOpts := token.GetUpdateOptions()
+	plannedTokenLabel := data.Label.ValueString()
+
+	if updateOpts.Label != plannedTokenLabel {
+		updateOpts.Label = plannedTokenLabel
+		token, err = client.UpdateToken(ctx, token.ID, updateOpts)
+
+		if err != nil {
+			resp.Diagnostics.AddError(
+				fmt.Sprintf("Failed to update the token with id %v", tokenID),
+				err.Error(),
+			)
+			return
+		}
```

Comment:
To generalize this logic a bit and prevent nesting, could we use a variable to track whether the resource has been updated and call `client.UpdateToken(...)` from within another if statement. This is a stylistic change so feel free to ignore 🙂

Reply:
Pushed.